diff -up tigervnc-1.1.0/common/network/TcpSocket.cxx.inetd-nowait tigervnc-1.1.0/common/network/TcpSocket.cxx
--- tigervnc-1.1.0/common/network/TcpSocket.cxx.inetd-nowait	2014-04-03 15:55:23.637911168 +0100
+++ tigervnc-1.1.0/common/network/TcpSocket.cxx	2014-04-03 15:57:03.144273441 +0100
@@ -308,18 +308,12 @@ bool TcpSocket::enableNagles(int sock, b
   return true;
 }
 
-bool TcpSocket::isSocket(int sock)
+bool TcpSocket::isListening(int sock)
 {
-  struct sockaddr_in info;
-  socklen_t info_size = sizeof(info);
-  return getsockname(sock, (struct sockaddr *)&info, &info_size) >= 0;
-}
-
-bool TcpSocket::isConnected(int sock)
-{
-  struct sockaddr_in info;
-  socklen_t info_size = sizeof(info);
-  return getpeername(sock, (struct sockaddr *)&info, &info_size) >= 0;
+  int listening = 0;
+  socklen_t listening_size = sizeof(listening);
+  return getsockopt(sock, SOL_SOCKET, SO_ACCEPTCONN, &listening,
+		    &listening_size) >= 0 && listening;
 }
 
 int TcpSocket::getSockPort(int sock)
diff -up tigervnc-1.1.0/common/network/TcpSocket.h.inetd-nowait tigervnc-1.1.0/common/network/TcpSocket.h
--- tigervnc-1.1.0/common/network/TcpSocket.h.inetd-nowait	2014-04-03 15:55:23.638911171 +0100
+++ tigervnc-1.1.0/common/network/TcpSocket.h	2014-04-03 15:56:03.880056590 +0100
@@ -56,8 +56,7 @@ namespace network {
     virtual void shutdown();
 
     static bool enableNagles(int sock, bool enable);
-    static bool isSocket(int sock);
-    static bool isConnected(int sock);
+    static bool isListening(int sock);
     static int getSockPort(int sock);
   private:
     bool closeFd;
diff -up tigervnc-1.1.0/unix/xserver/hw/vnc/vncExtInit.cc.inetd-nowait tigervnc-1.1.0/unix/xserver/hw/vnc/vncExtInit.cc
--- tigervnc-1.1.0/unix/xserver/hw/vnc/vncExtInit.cc.inetd-nowait	2014-04-03 15:55:23.597911024 +0100
+++ tigervnc-1.1.0/unix/xserver/hw/vnc/vncExtInit.cc	2014-04-03 15:55:23.638911171 +0100
@@ -220,8 +220,7 @@ void vncExtensionInit()
         network::TcpListener* listener = 0;
         network::TcpListener* httpListener = 0;
         if (scr == 0 && vncInetdSock != -1) {
-          if (network::TcpSocket::isSocket(vncInetdSock) &&
-              !network::TcpSocket::isConnected(vncInetdSock))
+          if (network::TcpSocket::isListening(vncInetdSock))
           {
             listener = new network::TcpListener(NULL, 0, 0, vncInetdSock, true);
             vlog.info("inetd wait");
