From e8d25d83463805c0f6ef623dba2ac9a276df3587 Mon Sep 17 00:00:00 2001
From: Luke Shumaker <lukeshu@lukeshu.com>
Date: Tue, 23 May 2017 02:16:27 -0400
Subject: vncviewer: Fix fullscreen scrolling


diff --git a/vncviewer/DesktopWindow.cxx b/vncviewer/DesktopWindow.cxx
index 47fe8f89..946b162c 100644
--- a/vncviewer/DesktopWindow.cxx
+++ b/vncviewer/DesktopWindow.cxx
@@ -1120,11 +1120,6 @@ void DesktopWindow::scrollTo(int x, int y)
   hscroll->value(x);
   vscroll->value(y);
 
-  if (!hscroll->visible())
-    x = -viewport->x();
-  if (!vscroll->visible())
-    y = -viewport->y();
-
   // Scrollbar position results in inverse movement of
   // the viewport widget
   x = -x;
@@ -1189,7 +1184,7 @@ void DesktopWindow::handleEdgeScroll(void *data)
   if ((dx == 0) && (dy == 0))
     return;
 
-  self->scrollTo(self->hscroll->value() + dx, self->vscroll->value() + dy);
+  self->scrollTo(self->hscroll->value() - dx, self->vscroll->value() - dy);
 
   Fl::repeat_timeout(0.1, handleEdgeScroll, data);
 }
