diff -up tigervnc-1.1.0/unix/xserver/hw/vnc/xf86vncModule.cc.extinit tigervnc-1.1.0/unix/xserver/hw/vnc/xf86vncModule.cc
--- tigervnc-1.1.0/unix/xserver/hw/vnc/xf86vncModule.cc.extinit	2014-07-15 16:10:41.793400143 +0100
+++ tigervnc-1.1.0/unix/xserver/hw/vnc/xf86vncModule.cc	2014-07-15 16:15:55.435004130 +0100
@@ -81,22 +81,27 @@ vncSetup(pointer module, pointer opts, i
 
 static void vncExtensionInitWithParams(INITARGS)
 {
-  rfb::initStdIOLoggers();
-  rfb::LogWriter::setLogParams("*:stderr:30");
-  rfb::Configuration::enableServerParams();
+  static char once = 0;
 
-  for (int scr = 0; scr < screenInfo.numScreens; scr++) {
-    ScrnInfoPtr pScrn = xf86Screens[scr];
+  if (!once) {
+    once++;
+    rfb::initStdIOLoggers();
+    rfb::LogWriter::setLogParams("*:stderr:30");
+    rfb::Configuration::enableServerParams();
 
-    for (ParameterIterator i(Configuration::global()); i.param; i.next()) {
-      const char *val;
+    for (int scr = 0; scr < screenInfo.numScreens; scr++) {
+      ScrnInfoPtr pScrn = xf86Screens[scr];
+
+      for (ParameterIterator i(Configuration::global()); i.param; i.next()) {
+	const char *val;
 #if XORG < 112
-      val = xf86FindOptionValue(pScrn->options, i.param->getName());
+	val = xf86FindOptionValue(pScrn->options, i.param->getName());
 #else
-      val = xf86FindOptionValue((XF86OptionPtr)pScrn->options, i.param->getName());
+	val = xf86FindOptionValue((XF86OptionPtr)pScrn->options, i.param->getName());
 #endif
-      if (val)
-        i.param->setParam(val);
+	if (val)
+	  i.param->setParam(val);
+      }
     }
   }
 
