Index: unix/xserver/hw/vnc/vncExtInit.cc
===================================================================
--- unix/xserver/hw/vnc/vncExtInit.cc	(revision 5013)
+++ unix/xserver/hw/vnc/vncExtInit.cc	(revision 5015)
@@ -591,13 +591,14 @@
     value1 = desktop1->getValueStr();
 
   /*
-   * Allow to change only clipboard parameters and desktop name.
+   * Allow to change only certain parameters.
    * Changing other parameters (for example PAM service name)
    * could have negative security impact.
    */
-  if (strcasecmp(param.buf, "desktop") != 0 &&
-      (noclipboard || strcasecmp(param.buf, "SendCutText") != 0) &&
-      (noclipboard || strcasecmp(param.buf, "AcceptCutText") != 0))
+  if (strncasecmp(param.buf, "desktop", 7) != 0 &&
+      strncasecmp(param.buf, "AcceptPointerEvents", 19) != 0 &&
+      (noclipboard || strncasecmp(param.buf, "SendCutText", 11) != 0) &&
+      (noclipboard || strncasecmp(param.buf, "AcceptCutText", 13) != 0))
     goto deny;
 
   rep.success = rfb::Configuration::setParam(param.buf);
